File: packages/codex-session-relay/src/codex_session_relay/hostadapter.py
"""What this package needs from a host, expressed as the smallest possible surface.

Every method is either a read or the one supported send. Nothing here can change a task's
model, effort, sandbox, approval policy, goal or archive state, because the relay has no
business doing any of that and an interface that cannot express it cannot do it by accident.
"""

from dataclasses import dataclass
from typing import Protocol


class HostUnavailable(Exception):
    """A read could not be completed. Not an answer, and never treated as one.

    Distinct from returning None: None means the host answered and the thing is absent, while
    this means we do not know. Conflating them is how an unreachable host becomes a false
    negative.
    """


@dataclass(frozen=True)
class TurnInfo:
    turn_id: str
    status: str
    started_at: float | None = None


@dataclass(frozen=True)
class TokenScan:
    """Result of looking for a delivery token in a recipient's items.

    The exhausted field is the important one. A scan that stopped at its bound has not
    searched the history, so not finding the token there means nothing at all.
    """

    found: bool
    turn_id: str | None
    exhausted: bool
    scanned: int


@dataclass(frozen=True)
class ThreadFacts:
    runtime_status: str
    can_accept_input: bool | None


# How far before the relay's own send stamp a turn's start may read and still be the turn that
# send started. The host reports whole seconds and the relay microseconds (ack.certainly_before
# allows the same second); the skew covers two clocks on one host. A turn that began earlier than
# this is older than the send, so a newest-first listing that reaches it has passed the place the
# dispatched turn would be.
TURN_START_PRECISION_SECONDS = 1.0
DISPATCH_TURN_SKEW_SECONDS = 60.0
TURN_PRESENT = "present"
TURN_ABSENT = "absent"
# How far back a dispatched-turn lookup pages before it gives up (1000 turns at the default page,
# ids and start times only). Checks run from 61 s after a send, so a parent has rarely begun more
# than a few turns by then; the bound is for a relay that was down while its parents kept working.
DISPATCHED_TURN_MAX_PAGES = 20


class ListingBounded(HostUnavailable):
    """The bounded listing never reached the send.

    Not evidence of absence, like any HostUnavailable - but unlike a failed read, waiting does not
    change it, so the relay records it by name instead of retrying it as if it were transient.
    """


@dataclass(frozen=True)
class TurnPresence:
    """What a recipient's own turn list says about the turn a dispatch started.

    present: the listing holds the turn. absent: the listing ended, or reached a turn that began
    before the send, without it; stop says which. Anything short of either is not an answer and
    raises HostUnavailable instead: an empty listing, or a bounded scan that never reached the
    send, has not shown that the turn is gone.

    seen names every listed turn read before the stop: the turns begun since the send, which is
    where the delivered message could be if it is anywhere.
    """

    finding: str
    turn: TurnInfo | None
    scanned: int
    stop: str
    seen: tuple = ()


def find_in_listing(pages, turn_id: str, sent_at: float) -> TurnPresence:
    """The one rule both adapters apply to a newest-first turn listing.

    pages yields (turns, another_page_follows), each page's TurnInfo newest first. The id is
    compared before the cutoff: a start that steered a turn begun before the send returns that
    turn's id, and that turn was the newest the thread had, so no older turn precedes it. A turn
    with no start time never serves as the cutoff.
    """
    cutoff = sent_at - TURN_START_PRECISION_SECONDS - DISPATCH_TURN_SKEW_SECONDS
    scanned = 0
    seen = []
    for turns, follows in pages:
        for turn in turns:
            scanned += 1
            if turn.turn_id == turn_id:
                return TurnPresence(TURN_PRESENT, turn, scanned, "matched", tuple(seen))
            if turn.started_at is not None and turn.started_at <= cutoff:
                return TurnPresence(TURN_ABSENT, None, scanned, "older_than_send", tuple(seen))
            seen.append(turn.turn_id)
        if not follows:
            if not scanned:
                raise HostUnavailable(
                    "the recipient's turn list is empty, which does not show that a turn is gone"
                )
            return TurnPresence(TURN_ABSENT, None, scanned, "listing_end", tuple(seen))
    raise ListingBounded(
        f"turn {turn_id!r} was not among {scanned} turns and the bounded listing never reached "
        "the send; this is not evidence of absence"
    )


def find_token_in(pages, token: str, turns) -> TokenScan:
    """A token among the items of the given turns, newest first: the rule both adapters apply.

    pages yields (items, another_page_follows), each item a (turn id, text) pair, newest first.
    exhausted means covered: the scan reached an item of a turn outside the set - a turn begun
    before the send - or the end of the items, so every item of those turns was read. A scan that
    stops at its bound first has not covered them, and not finding the token there shows nothing
    (I-42).
    """
    allowed = set(turns)
    scanned = 0
    for items, follows in pages:
        for owner, text in items:
            if owner is not None and owner not in allowed:
                return TokenScan(False, None, True, scanned)
            scanned += 1
            if token in text:
                return TokenScan(True, owner, False, scanned)
        if not follows:
            return TokenScan(False, None, True, scanned)
    return TokenScan(False, None, False, scanned)


class HostAdapter(Protocol):
    def read_thread(self, thread_id: str) -> ThreadFacts: ...

    def is_archived(self, thread_id: str, *, cwd: str | None = None) -> bool | None: ...

    def read_goal_status(self, thread_id: str) -> str | None: ...

    def list_turn_ids(self, thread_id: str, limit: int = 20) -> list: ...

    def read_turn(self, thread_id: str, turn_id: str) -> TurnInfo | None: ...

    def find_dispatched_turn(
        self, thread_id: str, turn_id: str, *, sent_at: float
    ) -> TurnPresence: ...

    def find_token_since(
        self, thread_id: str, token: str, *, turns, limit: int = 200
    ) -> TokenScan: ...

    def send_message(self, request_id: str, thread_id: str, message: str) -> dict: ...

    def get_operation(self, request_id: str) -> dict | None: ...

    def find_token(
        self, thread_id: str, token: str, *, limit: int = 200, turn_id: str | None = None
    ) -> TokenScan: ...

    def recipient_fingerprint(self, thread_id: str, *, window: int = 8) -> str: ...

File: internal/relay/delivery/hostadapter.go
package delivery

import (
	"fmt"
	"slices"
	"strings"
)

// Host reading rules shared by every adapter (hostadapter.py).
const (
	TurnStartPrecisionSeconds = 1.0
	DispatchTurnSkewSeconds   = 60.0
	TurnPresent               = "present"
	TurnAbsent                = "absent"
	InTurnItemsMax            = 2000
	userMessage               = "userMessage"
)

var agentOutput = []string{"agentMessage", "collabAgentToolCall", "commandExecution", "contextCompaction", "dynamicToolCall", "enteredReviewMode", "exitedReviewMode", "fileChange", "functionCallOutput", "imageGeneration", "imageView", "mcpToolCall", "plan", "reasoning", "sleep", "subAgentActivity", "webSearch"}

func isMessage(kind string) bool    { return kind == "" || kind == userMessage }
func mayBeMessage(kind string) bool { return !slices.Contains(agentOutput, kind) }

// ListingBounded: the bounded listing never reached the send (HostUnavailable).
type ListingBounded struct{ Message string }

func (e *ListingBounded) Error() string { return e.Message }

// ListingEmpty: the recipient listed no turns at all (HostUnavailable).
type ListingEmpty struct{ Message string }

func (e *ListingEmpty) Error() string { return e.Message }

// TurnPresence is hostadapter.TurnPresence.
type TurnPresence struct {
	Finding   string
	Turn      *TurnInfo
	Scanned   int
	Stop      string
	Seen      []any
	Older     []string
	SeenTurns []TurnInfo
	StopTurn  *TurnInfo
}

// ListingPage is one newest-first page of turns.
type ListingPage struct {
	Turns   []TurnInfo
	Follows bool
}

// Item is one thread item: (turn id, text, type); Type "" means untyped.
type Item struct{ Turn, Text, Type string }

// ItemPage is one page of items.
type ItemPage struct {
	Items   []Item
	Follows bool
}

func olderTurns(turns []TurnInfo, cutoff float64) []string {
	var out []string
	for _, t := range turns {
		if t.StartedAt != nil && *t.StartedAt <= cutoff {
			out = append(out, t.TurnID)
		}
	}
	return out
}

// FindInListing is find_in_listing over pages already read.
func FindInListing(pages []ListingPage, turnID string, sentAt float64) (TurnPresence, error) {
	next := 0
	return FindInListingPaged(func() (ListingPage, error) {
		page := pages[next]
		next++
		return page, nil
	}, len(pages), turnID, sentAt)
}

// FindInListingPaged is find_in_listing over a lazy listing: read yields the next page, at most
// bound pages are read, and nothing past the answer is read. turnID "" lists the turns since a
// send with no turn id; a listed turn without an id is never taken for it.
func FindInListingPaged(read func() (ListingPage, error), bound int, turnID string, sentAt float64) (TurnPresence, error) {
	cutoff := sentAt - TurnStartPrecisionSeconds - DispatchTurnSkewSeconds
	scanned := 0
	seen := []any{}
	var seenTurns []TurnInfo
	for p := 0; p < bound; p++ {
		page, err := read()
		if err != nil {
			return TurnPresence{}, err
		}
		for i, turn := range page.Turns {
			scanned++
			if turnID != "" && turn.TurnID == turnID {
				match := turn
				return TurnPresence{TurnPresent, &match, scanned, "matched", seen, olderAfterMatch(page.Turns[i+1:], page.Follows, read, bound-p-1, cutoff), seenTurns, nil}, nil
			}
			if turn.StartedAt != nil && *turn.StartedAt <= cutoff {
				stop := turn
				return TurnPresence{TurnAbsent, nil, scanned, "older_than_send", seen, olderTurns(page.Turns[i:], cutoff), seenTurns, &stop}, nil
			}
			seen = append(seen, idOf(turn))
			seenTurns = append(seenTurns, turn)
		}
		if !page.Follows {
			if scanned == 0 {
				return TurnPresence{}, &ListingEmpty{"the recipient's turn list is empty, which does not show that a turn is gone"}
			}
			return TurnPresence{TurnAbsent, nil, scanned, "listing_end", seen, nil, seenTurns, nil}, nil
		}
	}
	subject := "the send's turn"
	if turnID != "" {
		subject = "turn " + pyReprValue(turnID)
	}
	return TurnPresence{}, &ListingBounded{fmt.Sprintf("%s was not among %d turns and the bounded listing never reached the send; this is not evidence of absence", subject, scanned)}
}

// idOf is a listed turn's id, None when the host gave none.
func idOf(turn TurnInfo) any {
	if turn.TurnID == "" {
		return nil
	}
	return turn.TurnID
}

// olderAfterMatch is _older_after_match: after a match, read on to the first turn begun before
// the send. A failed read or the page bound gives none; the match stands either way.
func olderAfterMatch(rest []TurnInfo, follows bool, read func() (ListingPage, error), left int, cutoff float64) []string {
	for {
		for i, turn := range rest {
			if turn.StartedAt != nil && *turn.StartedAt <= cutoff {
				return olderTurns(rest[i:], cutoff)
			}
		}
		if !follows || left <= 0 {
			return nil
		}
		page, err := read()
		if err != nil {
			return nil
		}
		left--
		rest, follows = page.Turns, page.Follows
	}
}

// FindTokenIn is find_token_in: a token among a thread's items since a send, newest first.
func FindTokenIn(pages []ItemPage, token string, older []string) TokenScan {
	scanned := 0
	var other [2]any
	for _, page := range pages {
		for _, item := range page.Items {
			if item.Turn != "" && slices.Contains(older, item.Turn) {
				return TokenScan{false, nil, true, scanned, other[0], other[1]}
			}
			scanned++
			if !mayBeMessage(item.Type) || !strings.Contains(item.Text, token) {
				continue
			}
			if isMessage(item.Type) {
				return TokenScan{true, item.Turn, false, scanned, nil, nil}
			}
			if other[0] == nil && other[1] == nil {
				other = [2]any{item.Turn, item.Type}
			}
		}
		if !page.Follows {
			return TokenScan{false, nil, true, scanned, other[0], other[1]}
		}
	}
	return TokenScan{false, nil, false, scanned, other[0], other[1]}
}

// FindTokenInTurnItems is find_token_in_turn_items.
func FindTokenInTurnItems(pages []ItemPage, token, turnID string) TokenScan {
	scanned := 0
	var other [2]any
	for _, page := range pages {
		for _, item := range page.Items {
			scanned++
			if item.Turn != turnID || !mayBeMessage(item.Type) || !strings.Contains(item.Text, token) {
				continue
			}
			if isMessage(item.Type) {
				return TokenScan{true, item.Turn, false, scanned, nil, nil}
			}
			if other[0] == nil && other[1] == nil {
				other = [2]any{item.Turn, item.Type}
			}
		}
		if !page.Follows {
			return TokenScan{false, nil, true, scanned, other[0], other[1]}
		}
	}
	return TokenScan{false, nil, false, scanned, other[0], other[1]}
}
